### PR TITLE
Parser no longer takes greedy. Accepts indirect selection, a bool.

### LINKED ISSUE #4082: Return to eager test selection by default, with an option to tone it down
### Is there an existing feature request for this?

- [X] I have searched the existing issues

### Describe the Feature

- dbt 0.20.0 changed how tests were selected, to avoid errors in more complex dbt projects that don't build the entire project during development/CI.
    - Specifically: In a fresh schema, operators like `@` and `+` can cause models to be tested that don't exist ([1827](https://github.com/dbt-labs/dbt-core/issues/1827), [2132](https://github.com/dbt-labs/dbt-core/issues/2132)). Singular tests that touch a model are run even if they are only indirectly selected ([3832](https://github.com/dbt-labs/dbt-core/issues/3382))
    - The change was to only run tests if *all* first-order parents are included in selection ([2891](https://github.com/dbt-labs/dbt-core/issues/2891)). The writeup in the PR ([3235](https://github.com/dbt-labs/dbt-core/pull/3235)) is very good so I won't re-summarise it.
- As an unfortunate side effect, it broke people's mental model (it used to do what you *meant*; now it does what you asked)
- I was originally writing an article explaining why the changes are important and helping dbt users understand when they should use the new-in-0.21.0 `--greedy` flag. As I dug into the edge cases, my thinking changed and **I now think we should restore the 0.19.0 behaviour as the default**, with an option to disable greedy selection for the use cases that necessitated the changes in the first place. 

Matrix of functionality/options:
image

### Describe alternatives you've considered

- Make the new behaviour much more discoverable.
    - I think [if the DAG in dbt Docs also visualised tests](https://getdbt.slack.com/archives/C2JRTGLLS/p1633962686096200), then we could more reasonably expect users to preview their selections there (I used to use it as a visual debugger of complex selection syntax)
    - Document the crap out of it, particularly in the tutorial.
    
- Change nothing: there's a risk of creating *more* confusion by flipping back to the old way and annoying the people who liked it
    - If we're going to change, we should really do it before 1.0 or we're stuck.
    - There's already significant confusion in the community, for both new and longstanding members. [3496](https://github.com/dbt-labs/dbt/issues/3496), [3571](https://github.com/dbt-labs/dbt/issues/3571), [3706](https://github.com/dbt-labs/dbt/issues/3706), [3903](https://github.com/dbt-labs/dbt/issues/3903), [Slack 1](https://getdbt.slack.com/archives/C01UM2N7814/p1633623002202700) (9mths in community), [Slack 2](https://getdbt.slack.com/archives/C01UM2N7814/p1632336696127500) (2yrs in community), [Slack 3](https://getdbt.slack.com/archives/C01UM2N7814/p1633425377185100) (5mths in community), [Slack 4](https://getdbt.slack.com/archives/C2JRRQDTL/p1626271497173000) (18mths in community), [Slack 5](https://getdbt.slack.com/archives/C01UM2N7814/p1630508257048800) (new to community), [Slack 6](https://getdbt.slack.com/archives/CBSQTAPLG/p1630096188152800) (1mth in community)
- Try harder to internalise the new system.
    - I don't think it's better to expect this of new users: the old failure modes were real, but they tended to impact advanced dbt users (the green users). 
    image
    _NB: Venn diagram is entirely unscientific_
    - The new failure mode catches people immediately (the yellow users), especially since the [Tutorial](https://docs.getdbt.com/tutorial/test-and-document-your-project) currently says "Running tests on one model looks very similar to running a model: use the `--select` flag (or `-s` flag), followed by the name of the model"

### Who will this benefit?

- People whose mental model has been broken by the changes
    - Specifically, who expect to be able to switch between `dbt run -m some_model` and `dbt test -m some_model` and have all tests attached to `some_model` run.
    - The union of `dbt test --exclude tag:x` and `dbt test --select tag:x` is no longer the same as `dbt test` when one model in a test is tagged `x` and one isn't.
    - Side note: [docs#514](https://github.com/dbt-labs/docs.getdbt.com/issues/514) correctly calls for endless examples of node selection. We should do this, no matter what the outcome here!
- A lot of the takeaway from my original post was going to be "the good thing is that there's now a flag to allow you to choose the right approach for your use case". That remains true even if we flip the default back.
- I wanted to be able to provide a clear set of guidelines covering when you should use `--greedy` and not. All of the cases devolved towards "yes, but..." or "it depends". If the settings were flipped,  I think a lot of the guidance becomes easier because we can describe the handful of cases where it becomes desirable with less ambiguity: "Exclude the other parents when they won't exist"

### Are you interested in contributing this feature?

Sure am! @VersusFacit and I are going to tag-team it 

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by VersusFacit**
resolves #4082 

### Description

This removes the greedy flag and instead adds the indirect-selection flag with eager and cautious as options for which files to pull in as nodes when running tests.

### Checklist

- [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
- [x] I have run this code in development and it appears to resolve the stated issue
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [x] I have updated the `CHANGELOG.md` and added information about my change

### CODE REVIEW

File: test/integration/066_test_selection_tests/test_selection_expansion.py [hunk 1]
```diff
@@ -225,84 +258,94 @@ def test__postgres__test_level_tag(self):
...
                 value: model_a
-                greedy: false
-            - name: model_a_greedy_true
+                indirect_selection: "cautious"
+            - name: model_a_yes_indirect_selection
               definition:
                 method: fqn
                 value: model_a
-                greedy: true
+                indirect_selection: "eager"
```
[the remark this answers is not in the document]
Comment (VersusFacit) [reply test/integration/066_test_selection_tests/test_selection_expansion.py:328]: I found a way (see core/dbt/graph/selector_spec.py) after tracing around the runtime for a nontrivial amount of time 😅 

I wrote the code so it ignores command line args if the selector has the field. A runtime exception is thrown if the selector is malformed, that is, something other than eager/cautious is in the yaml.

**Comment by VersusFacit** [comment]
@jtcohen6  Burnt some midnight oil and figured out a way to solve this. Tests are working and I added some. The code allows users to add the field to a selector yaml and override whatever the CLI flag may or may not be.

Do you want to mix in the changes in that linked sha? Is the total renaming necessary? I almost like the way the predicates flow but you know the structure here better than I.

**Review by jtcohen6 — APPROVED** [review]
Thanks for the hard work to see this through!

As long as the end-user behavior is right, and the argument/property names make enough sense to follow through the code, this is as much as we need for now. At such time as we want to swing back through and add a third (or fourth) indirect selection mode, we can lightly refactor as it makes sense. I don't think we need to spend any more time here now.

@VersusFacit As far as the changelog: Could you add a line under the `dbt-core 1.0.0 (Release TBD)` > `Features`, with a link to the issue + this PR? You can also add yourself + Joel to the list of contributors. As soon as you do that, we can merge and lock this in for v1 :)

After that, there'll just be the matter of updating docs: https://github.com/dbt-labs/docs.getdbt.com/issues/896

**Comment by jtcohen6** [comment]
I want to be sure this change gets in before v1 (especially the switch to the default), so I'm going to add a changelog entry now and merge!

**Pull request merged**